nano/secure/ledger_set_confirmed.cpp

```diff
@@ -0,0 +1,161 @@
+#include <nano/store/account.hpp>
+#include <nano/store/component.hpp>
+#include <nano/store/confirmation_height.hpp>
+#include <nano/store/pending.hpp>
+#include <nano/store/pruned.hpp>
+
+nano::ledger_set_confirmed::ledger_set_confirmed (nano::ledger const & ledger) :
+	ledger{ ledger }
+{
+}
+
+std::optional<nano::uint128_t> nano::ledger_set_confirmed::balance (store::transaction const & transaction, nano::account const & account_a) const
+{
+	auto block = get (transaction, head (transaction, account_a));
+	if (!block)
+	{
+		return std::nullopt;
+	}
+	return block->balance ().number ();
+}
+
+// Balance for account containing hash
+std::optional<nano::uint128_t> nano::ledger_set_confirmed::balance (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	if (hash.is_zero ())
+	{
+		return std::nullopt;
+	}
+	auto block = get (transaction, hash);
+	if (!block)
+	{
+		return std::nullopt;
+	}
+	return block->balance ().number ();
+}
+
+bool nano::ledger_set_confirmed::exists (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	auto block = ledger.store.block.get (transaction, hash);
+	if (!block)
+	{
+		return false;
+	}
+	auto info = ledger.store.confirmation_height.get (transaction, block->account ());
+	if (!info)
+	{
+		return false;
+	}
+	return block->sideband ().height <= info.value ().height;
+}
+
+bool nano::ledger_set_confirmed::exists_or_pruned (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	if (ledger.store.pruned.exists (transaction, hash))
+	{
+		return true;
+	}
+	auto block = ledger.store.block.get (transaction, hash);
+	if (!block)
+	{
+		return false;
+	}
+	auto info = ledger.store.confirmation_height.get (transaction, block->account ());
+	if (!info)
+	{
+		return false;
+	}
+	return block->sideband ().height <= info.value ().height;
+}
+
+std::optional<nano::account_info> nano::ledger_set_confirmed::get (store::transaction const & transaction, nano::account const & account) const
+{
+	auto info = ledger.store.confirmation_height.get (transaction, account);
+	if (!info)
+	{
+		return std::nullopt;
+	}
+	debug_assert (false);
+	return ledger.store.account.get (transaction, account);
+}
+
+std::shared_ptr<nano::block> nano::ledger_set_confirmed::get (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	auto block = ledger.store.block.get (transaction, hash);
+	if (!block)
+	{
+		return nullptr;
+	}
+	auto info = ledger.store.confirmation_height.get (transaction, block->account ());
+	if (!info)
+	{
+		return nullptr;
+	}
+	return block->sideband ().height <= info.value ().height ? block : nullptr;
+}
+
+std::optional<nano::pending_info> nano::ledger_set_confirmed::get (store::transaction const & transaction, nano::pending_key const & key) const
+{
+	auto result = ledger.store.pending.get (transaction, key);
+	if (!result && !exists_or_pruned (transaction, key.hash))
```

Comment:
Is this expression right? it looks wrong to me I find it very difficult to make sense of. Splitting into if statements would improve readability. It seems always return pending entry if it exists.

Reply:
It wasn't but the code also wasn't used so I removed it.